File: 2023/d03.R
parse_input <- function(path) {
  d <- readLines(path)
  n <- nchar(d[[1]]) + 1L
  s <- paste(d, collapse = "\n")
  start <- gregexpr("([0-9]+)", s)[[1]] - 1L
  end <- start + attr(start, "match.length") - 1L
  value <- as.integer(substr(rep(s, length(start)), start + 1, end + 1))
  numbers <- cbind(row = start %/% n + 1,
                   col0 = start %% n + 1,
                   col1 = end %% n + 1)
  i <- gregexpr("[^.0-9\n]", s)[[1]] - 1L
  symbols <- cbind(row = i %/% n + 1,
                   col = i %% n + 1)

  near <- abs(outer(numbers[, 1], symbols[, 1], "-")) < 2 &
    (abs(outer(numbers[, 2], symbols[, 2], "-")) < 2 |
     abs(outer(numbers[, 3], symbols[, 2], "-")) < 2)
  colnames(near) <- substr(rep(s, length(i)), i + 1, i + 1)

  list(numbers = value, near = near)
}


part1 <- function(d) {
  sum(d$numbers[apply(d$near, 1, any)])
}


part2 <- function(d) {
  near <- d$near[, colnames(d$near) == "*"]
  sum(apply(near, 2, function(i) if (sum(i) == 2) prod(d$numbers[i]) else 0))
}

test <- parse_input("test-input-d03.txt")
stopifnot(part1(test) == 4361)
stopifnot(part2(test) == 467835)

d <- parse_input("input-d03.txt")
part1(d)
part2(d)

File: 2023/d03.py
import re

def parse_input(path):
    with open(path) as f:
        d = f.read()
    n = d.find("\n") + 1
    numbers = []
    for it in re.finditer("([0-9]+)", d):
        start, end = it.span()
        numbers.append((start // n, start % n, end % n, int(d[start:end])))
    symbols = [[] for _ in range(n - 1)]
    for it in re.finditer("[^.0-9\n]", d):
        start = it.start()
        symbols[start // n].append((start % n, d[start]))
    return numbers, symbols


def part1(d):
    def test_number(x):
        row, col_start, col_end, value = x
        for r in range(max(0, row - 1), min(len(symbols), row + 2)):
            for c, _ in symbols[r]:
                if c >= col_start - 1 and c <= col_end:
                    return value
        return 0
    numbers, symbols = d
    return sum([test_number(x) for x in numbers])


# This is a bit gross as we loop through all the numbers for every "*"
# but it could be worse, really.
def part2(d):
    def test_times(r, c):
        pos = []
        for num in numbers:
            row, col_start, col_end, value = num
            if abs(row - r) <= 1 and c >= col_start - 1 and c <= col_end:
                pos.append(value)
        return pos[0] * pos[1] if len(pos) == 2 else 0
    tot = 0
    numbers, symbols = d
    for r in range(len(symbols)):
        for c, v in symbols[r]:
            if v == "*":
                tot += test_times(r, c)
    return tot


d = parse_input("test-input-d03.txt")
assert part1(d) == 4361
assert part2(d) == 467835

d = parse_input("input-d03.txt")
print(part1(d))
print(part2(d))
